src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -195,71 +232,78 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
             if (targetStatement.IsKind(SyntaxKind.IfStatement, out IfStatementSyntax ifStatement))
             {
                 CheckDefiniteAssignment(
-                    semanticModel, localVariable, ifStatement.Else,
-                    out isAssigned, out isAccessedBeforeAssignment,
-                    cancellationToken);
+                    semanticModel,
+                    localVariable,
+                    isNegativeNullCheck ? ifStatement.Statement : ifStatement.Else?.Statement, lastStatement: null,
+                    out isAssigned, 
+                    out isAccessedBeforeAssignment);
 
                 if (isAccessedBeforeAssignment)
                 {
-                    return true;
+                    return false;
                 }
 
                 if (isAssigned)
                 {
-                    return false;
+                    return true;
                 }
             }
 
             // Make sure that no access is made to the variable before assignment in the subsequent statements
-            for (int index = targetIndex + 1, n = statements.Count; index < n; index++)
+            if (targetIndex + 1 < statements.Count)
             {
                 CheckDefiniteAssignment(
-                    semanticModel, localVariable, statements[index],
-                    out isAssigned, out isAccessedBeforeAssignment,
-                    cancellationToken);
-
-                if (isAccessedBeforeAssignment)
+                    semanticModel, 
+                    localVariable,
+                    statements[targetIndex + 1],
+                    statements[statements.Count - 1],
+                    out isAssigned,
+                    out isAccessedBeforeAssignment);
+
+                if (isAccessedBeforeAssignment || isAssigned)
                 {
-                    return true;
-                }
+                    if (targetStatement.Kind() == SyntaxKind.WhileStatement)
+                    {
+                        // The scope of pattern variables in a while-statement does not leak out to
+                        // the enclosing block so we bail also if there is any assignments afterwards.
+                        return false;
+                    }
 
-                if (isAssigned)
-                {
-                    // The scope of pattern variables in a while-statement does not leak out to
-                    // the enclosing block so we bail also if there is any assignments afterwards.
-                    return targetStatement.Kind() == SyntaxKind.WhileStatement;
+                    if (isAssigned)
+                    {
+                        return true;
+                    }
+
+                    if (isAccessedBeforeAssignment)
+                    {
+                        // Access before assignment here is only valid if we have a negative
+                        // pattern-matching in an if-statement with an unreachable endpoint.
```

Comment:
💡 Assuming my understanding is correct, the following might be good to add.

```csharp
// For example:
//
// if (!(o is string x)) {
//   return;
// }
//
// // The 'return' statement above ensures x is definitely assigned here
// Console.WriteLine(x);
```

Reply:
Done.